### PR TITLE
test(add tests): unit and integration tests

### PR DESCRIPTION
**Pull request by nemni8**
WIP
----------------------------

this PR is changing the way public, backend and pages get their declarations.
I splitted the declaration.d.ts file into 3 files.
one that includes node module typings (Buffer)
one that includes $w typings
and all other declarations
currently i have 3 green basic tests basic that compile sample-code from all types (backend, public and pages).

### CODE REVIEW

File: configs/tsconfig.backend.json [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
   "extends": "./tsconfig.base.json",
   "compilerOptions": {
-    "types": ["corvid-types/types/backend"]
+    "typeRoots": ["../", "node_modules/@types"],
+    "types": ["types/backend", "node"]
```
Comment (dor-itzhaki) [inline configs/tsconfig.backend.json]: please move the changes to the 3 config files to a different commit explaining what the change does

File: package.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
     "lint": "eslint .",
     "build": ":",
     "test": "jest",
-    "generate-dts": "docworks dts -r git@github.com:wix-private/wix-code-docs.git -o declaration -d temp_declarations --wixselector && npm run copy-declarations temp_declarations",
+    "generate-dts": "docworks dts -r git@github.com:wix-private/wix-code-docs.git -b captcha-fix -o declaration -d temp_declarations --wixselector && npm run copy-declarations temp_declarations",
```
Comment (dor-itzhaki) [inline package.json]: is this intentional ?

File: package.json [hunk 3]
```diff
@@ -51,5 +55,9 @@
     "*.js": [
       "eslint"
     ]
+  },
+  "dependencies": {
+    "@types/node": "^12.7.5",
+    "tslint": "^5.20.0"
```
Comment (dor-itzhaki) [inline package.json]: should `tslint` be a dev dependency ?

File: test/it/negative.spec.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+const compiler = require("../utils/compiler");
+const testRoots = require("../utils/tests-roots");
+const negativeRoots = testRoots.getNegativeRoots();
+
+describe("typescript - negative scenarios", () => {
+  it.each(negativeRoots)(
+    "should NOT complie successfully %s folder",
```
Comment (dor-itzhaki) [inline test/it/negative.spec.js]: "should fail compiling $s"

File: test/it/negative.spec.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    async rootPath => {
+      const configurationFilename = `${testRoots.NEGATIVE_ROOT_PATH}/${rootPath}/tsconfig.json`;
+      const result = compiler(configurationFilename);
```
Comment (dor-itzhaki) [inline test/it/negative.spec.js]: i think it would br nicer if in case of an error, the compiler function will throw with the relevant errors.
if you prefer the current api, let's at least rename "result" to "errors"

File: test/it/positive.spec.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+const compiler = require("../utils/compiler");
+const testRoots = require("../utils/tests-roots");
+const positiveRoots = testRoots.getPositiveRoots();
+
+describe("typescript - positive scenarios", () => {
+  it.each(positiveRoots)(
+    "should complie successfully %s folder",
```
Comment (dor-itzhaki) [inline test/it/positive.spec.js]: "should successfully compile %s"

File: types/backend/index.d.ts [hunk 7]
```diff
@@ -1 +1,4 @@
-/// <reference path="../common/declaration.d.ts" />
\ No newline at end of file
+/// <reference path="../common/declaration.d.ts" />
```
Comment (dor-itzhaki) [inline types/backend/index.d.ts]: can you move the changes regarding $w compilation errors to a different commit ?

File: types/common/declaration.d.ts [hunk 8]
```diff
@@ -2853,7 +2853,7 @@ declare namespace $w {
         /**
          * A verification event handler.
          */
-        type VerifiedHandler = ()=>Promise<Void>;
+        type VerifiedHandler = ()=>Promise<void>;
```
Comment (dor-itzhaki) [inline types/common/declaration.d.ts]: was this automatically generated ?

File: types/common/buffer.d.ts [hunk 9]
```diff
@@ -0,0 +1,232 @@
+declare namespace NodeJS {
```
Comment (dor-itzhaki) [inline types/common/buffer.d.ts]: can we just "import" the node buffer definition instead of copying it here ?
(we anyway depend on node types)

File: test/utils/tests-roots.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+const { readdirSync } = require("fs");
```
Comment (dor-itzhaki) [inline test/utils/tests-roots.js]: i don't think this file is necessary.
you can have a file for "listSubDirectories" maybe, as a utility.
then inside negative.spec for example, just do:
`it.each(listSubDirectories("./code-smaples/negative"))...`

File: test/utils/compiler.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+  );
+
+  const { options, fileNames } = parsedCmd;
+
+  const program = ts.createProgram({
+    rootNames: fileNames,
+    options
+  });
+
+  console.log(`compiling ${fileNames.length} files`, fileNames);
```
Comment (dor-itzhaki) [inline test/utils/compiler.js]: remove ?

File: test/utils/compiler.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+          line,
+          character
+        } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
+        msg = `${diagnostic.file.fileName} (${line + 1},${character +
+          1}): ${msg}`;
+      }
+      compilerErrors.push(msg);
+    });
+
+  return compilerErrors;
```
Comment (dor-itzhaki) [inline test/utils/compiler.js]: as i wrote above... consider throwing in case of errors

File: test/it/negative.spec.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+      expect(result.length).toBeGreaterThan(0);
```
Comment (dor-itzhaki) [inline test/it/negative.spec.js]: i am missing the reason for failure. if the reason is not what we expect, the test doesn't cover what we think it covers.
maybe we can add a ".expectedErrors.txt" file (or something similar) to each directory and compare it to the errors that the compiler returns for each test.

**Comment by dor-itzhaki** [comment]
you should keep the negative test code examples as minimal as possible so it will be easier to understand what they test.
for example, in the `negative/backend-$w-module`, keep only the `$w-module.js` file which contains the tested error.

regarding the positive code examples... we should later on think on adding much more content to them

File: package.json (hunk 2 above)
Comment (nemni8) [reply package.json]: no,fixed

File: package.json [hunk 14]
```diff
@@ -36,10 +36,14 @@
     "eslint": "^6.3.0",
     "eslint-config-prettier": "^6.3.0",
     "eslint-plugin-import": "^2.18.2",
+    "eslint-plugin-jest": "^22.17.0",
```
Comment (dor-itzhaki) [inline package.json]: sorry to be picky, but the new devDependencies and the rename of the index.js file, don't belong to this commit

File: test/it/positive.spec.js [hunk 15]
```diff
@@ -0,0 +1,15 @@
+const compiler = require("../utils/compiler");
+const listSubDirectories = require("../utils/listSubDirectories");
+const POSITIVE_ROOT_PATH = "test/it/code-samples/positive";
+const positiveRoots = listSubDirectories(POSITIVE_ROOT_PATH);
+
+describe("typescript - positive scenarios", () => {
+  it.each(positiveRoots)(
+    "should successfully compile %s folder",
+    async tsRootPath => {
+      const configurationFilename = `${POSITIVE_ROOT_PATH}/${tsRootPath}/tsconfig.json`;
```
Comment (dor-itzhaki) [inline test/it/positive.spec.js]: if "listSubDirectories" would just concat the subdir to the root dir, you will not need to "join" here

File: test/utils/compiler.js [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
+        msg = `${getFileName(diagnostic.file.fileName)} (${line +
+          1},${character + 1}): ${msg}`;
+      }
+      compilerErrors.push(msg);
+    });
+
+  if (compilerErrors.length > 0) throw new Error(compilerErrors.toString());
+
+  return compilerErrors;
```
Comment (dor-itzhaki) [inline test/utils/compiler.js]: just don't return anything in case of an error

File: scripts/copy_declarations.sh [hunk 17]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+TEMP_DIR=$1
+
+echo "copying declaration.d.ts to types/common/declaration.d.ts"
```
Comment (tomenden) [inline scripts/copy_declarations.sh]: Why do we need this script? Can we provide docworks with the dir we want from the start?
i.e, change:
https://github.com/wix-incubator/corvid-types/blob/7972d8e4f7414216d13b6f2198b05375814b906e/package.json#L9
to `-d types/common`

File: types/backend/index.d.ts [hunk 18]
```diff
@@ -1 +1,4 @@
...
+declare namespace $w {
```
Comment (tomenden) [inline types/backend/index.d.ts]: I remember we saw that adding the namespace on the backend had some undesired side-effects (such as suggesting $w.$w function). Can we improve it? Can you remind me why we need this definition?

File: types/pages/index.d.ts [hunk 19]
```diff
@@ -1,2 +1,5 @@
+/// <reference path="../common/$w.d.ts" />
 /// <reference path="../common/declaration.d.ts" />
-/// <reference path="./$w.d.ts" />
\ No newline at end of file
+
+declare class Buffer {
```
Comment (tomenden) [inline types/pages/index.d.ts]: Can you explain why pages need a Buffer?
Worth documenting these things, because they are not self-explanatory in code..
Consider either comments (though in dts they might have undesired side effects), a markdown file, or github issues

File: types/public/index.d.ts [hunk 20]
```diff
@@ -1 +1,6 @@
-/// <reference path="../common/declaration.d.ts" />
\ No newline at end of file
+/// <reference path="../common/$w.d.ts" />
+/// <reference path="../common/declaration.d.ts" />
+declare class Buffer {
+}
+
+type PageElementsMap = {}
```
Comment (tomenden) [inline types/public/index.d.ts]: Why do we need the initial empty declaration?
